### PR TITLE
vmware: Make sure instance memory is multiple of 4.

### PR DESCRIPTION
**Pull request by kpawar-sap**

### CODE REVIEW

**Review by joker-at-work — CHANGES_REQUESTED** [review]
I don't think it makes sense to put that as a new capability for every driver, unless you discussed this with upstream as the way to go. Instead, I would push the main code (`_validate_memory_size()`) into the vmware driver and let it raise the `BuildAbortException` instead. This would also contain the changes to the `nova/virt/vmwareapi` directory and thus the changes are less likely to clash with upstream changes.

**Comment by joker-at-work** [comment]
> Possibly a change in the RAM filter

I wouldn't got there. This would just mean that no host is reported as compatible without a clear error message. Having the vmware driver raise an explicit message is imho preferrable, as I would assume that message also bubbles up into the tickets customers will send us and thus lessen the support effort.

**Comment by joker-at-work** [comment]
This would also limit flavor-creation for future KVM and for current baremetal hosts. I wouldn't push it there. Only the vmware hypervisor has - to my knowledge - this problem.

**Comment by fwiesel** [comment]
> This would also limit flavor-creation for future KVM and for current baremetal hosts. I wouldn't push it there. Only the vmware hypervisor has - to my knowledge - this restriction.

Let's be clear about the goal/scope. I only want a solution addressing our need to reduce the support workload with our current configuration. Both proposals need work if we go for KVM or other hypervisors.

Our Baremetal hardware has the same constraint, since the smallest DRAM bank size increase is exactly 4MB.
See the [SDP](https://en.wikipedia.org/wiki/Serial_presence_detect#SDR_SDRAM), byte 0x1f. Probably the reason VMWare has the constraint.

While it may be possible to configure different RAM sizes, I do not really understand the necessity. For baremetal it is purely accounting Still, if the need arise, we still can revisit the problem again, and replace it with something else.

But should we find out that other hypervisors have similar constraints (since they are emulating hardware, which has it), we might even be able to upstream it. But I was considering that out of scope for now.

File: nova/exception.py [hunk 1]
```diff
@@ -1841,6 +1841,11 @@ class RequestedVRamTooHigh(NovaException):
                 "than the maximum allowed by flavor %(max_vram)d.")
 
 
+class RequestedRamSizeUnsupported(NovaException):
+    msg_fmt = _("The requested amount of memory %(req_ram)d is not supported "
+                "memory size.")
```
Comment (joker-at-work) [inline nova/exception.py:1846]: I think there's a way to provide additional information into the exception message. The current way is too generic for me. I'd like to have the "not a multiple of 4" in there. There are other exceptions that allow the raising code to pass an additional "reason" iirc.

**Comment by fwiesel** [comment]
I would suggest to do the check in the pre-flight check [prepare_for_spawn](https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/virt/driver.py#L283-L293). It seems to fit the description of the interface, and would happen before the actual spawn process. So no network attachments, volume creations, etc...)

**Comment by kpsap** [comment]
> I would suggest to do the check in the pre-flight check [prepare_for_spawn](https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/virt/driver.py#L283-L293). It seems to fit the description of the interface, and would happen before the actual spawn process. So no network attachments, volume creations, etc...)
> 
> Clarification: That means the method needs to be overridden in the vmwareapi driver and filled with the check

Fixed

**Comment by kpawar-sap** [comment]
> Sorry, one more thing. If I see it correctly, the exception drivers (mostly hyperv) seem to raise for an invalid configuration is `InstanceUnacceptable`, I would prefer that one over our own custom one.

Fixed

**Review by fwiesel — APPROVED** [review]

File: nova/virt/vmwareapi/driver.py [hunk 2]
```diff
@@ -533,6 +533,10 @@ def get_inventory(self, nodename):
                 }
         return result
 
+    def prepare_for_spawn(self, instance):
+        """Prepare to spawn instance."""
```
Comment (joker-at-work) [inline nova/virt/vmwareapi/driver.py:537]: This comment doesn't give any information the name of the function doesn't tell already. We could say, we do pre-checks here or just leave it out.

Comment (kpawar-sap) [reply nova/virt/vmwareapi/driver.py]: Fixed

**Review by joker-at-work — APPROVED** [review]

**Pull request merged**